doc/source/ray-core/examples/rdt/grpo_contextual_bandits.py

```diff
@@ -0,0 +1,432 @@
+        advantages: torch.Tensor,
+    ) -> torch.Tensor:
+        """Apply PPO update to policy network."""
+        # Compute new log probabilities and ratio
+        dist_new = Categorical(logits=self.model(states))
+        new_logps = dist_new.log_prob(actions)
+        ratio = (new_logps - old_logps).exp()
+        unclipped = ratio * advantages
+        clipped = torch.clamp(ratio, 1 - PPO_CLIP_EPS, 1 + PPO_CLIP_EPS) * advantages
+        ppo_loss = -torch.min(unclipped, clipped).mean()
+
+        # Compute KL divergence between new policy and reference policy
+        with torch.no_grad():
+            ref_logits = self.ref_model(states)
+        kl = kl_divergence(Categorical(logits=ref_logits), dist_new).mean()
+
+        # Combine PPO loss and KL divergence to prevent large policy updates
+        loss = ppo_loss + KL_COEFF * kl
+
+        # Update policy network
+        self.optim.zero_grad()
+        loss.backward()
+        nn.utils.clip_grad_norm_(self.model.parameters(), GRAD_CLIP_NORM)
+        self.optim.step()
+        # Update EMA teacher weights
+        with torch.no_grad():
+            for p_ref, p in zip(self.ref_model.parameters(), self.model.parameters()):
+                # decay * old + (1 - decay) * new
+                p_ref.mul_(EMA_DECAY).add_(p.data, alpha=1.0 - EMA_DECAY)
+        self.policy_version += 1
+        return loss.detach()
+
+    def step(self) -> dict[str, Any]:
+        """Perform one training step and return lightweight metrics."""
+        slices: list[TrajectorySlice] = ray.get(
+            self.replay_buffer.sample_from.remote(BATCH_SIZE)
+        )
+        raw_states = torch.stack([s["state"] for s in slices])
+        actions = torch.cat([s["actions"] for s in slices])
+        old_logps = torch.cat([s["old_logps"] for s in slices])
+        rewards = torch.cat([s["rewards"] for s in slices])
+
+        # Track cosine gap between best possible action and model prediction for reporting.
+        first_state = raw_states[0]
+        first_reward = float(rewards[0].item())
+        cosine_values = torch.mv(action_dirs, first_state)
+        best_first_reward = float(torch.max(cosine_values).item())
+        cosine_gap = abs(best_first_reward - first_reward)
+
+        # Prepare tensors for update
+        states = raw_states.repeat_interleave(ACTION_DIM, 0).to("cuda")
+        actions = actions.to("cuda")
+        old_logps = old_logps.to("cuda")
+        rewards = rewards.to("cuda")
+
+        # Compute advantages and update policy
+        advs = self._compute_advantages(rewards)
+        loss = self._apply_policy_update(states, actions, old_logps, advs)
+
+        return {
+            "loss": float(loss.item()),
+            "cosine_gap": float(cosine_gap),
+        }
+
+    # NOTE: cannot use async method with tensor transport
+    @ray.method(tensor_transport="nixl")
+    def get_weights(self) -> dict[str, torch.Tensor]:
+        """Return a GPU ObjectRef to current model weights."""
```

Comment:
```suggestion
        """The tensor_transport="nixl" option uses NIXL via RDT to transfer model weight tensors. Removing it will default to the Ray object store."""
```